### ISSUE TITLE
sstable_compaction_test slows down terribly under --overprovisioned

### ISSUE THREAD

**Issue by xemul** [scylladb/scylladb#12795]
The `--overprovisioned` is how test.py runs the guy and it becomes terribly slow

```
$ time ./build/dev/test/boost/sstable_compaction_test --run_test=twcs_reshape_with_disjoint_set_test -- --random-seed=2862905288 -c1 -m2G 
Running 1 test case...
INFO  2023-02-09 18:36:57,459 seastar - Reactor backend: linux-aio
WARN  2023-02-09 18:36:57,459 [shard 0] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
INFO  2023-02-09 18:36:57,460 [shard 0] seastar - Created fair group io-queue-0, capacity rate 2147483:2147483, limit 12582912, rate 16777216 (factor 1), threshold 2000
INFO  2023-02-09 18:36:57,460 [shard 0] seastar - IO queue uses 0.75ms latency goal for device 0
INFO  2023-02-09 18:36:57,460 [shard 0] seastar - Created io group dev(0), length limit 4194304:4194304, rate 2147483647:2147483647
INFO  2023-02-09 18:36:57,460 [shard 0] seastar - Created io queue dev(0) capacities: 512:2000:2000 1024:3000:3000 2048:5000:5000 4096:9000:9000 8192:17000:17000 16384:33000:33000 32768:65000:65000 65536:129000:129000 131072:257000:257000
random-seed=2862905288

*** No errors detected

real	0m1,400s
user	0m0,536s
sys	0m0,699s
```

```
$ time ./build/dev/test/boost/sstable_compaction_test --run_test=twcs_reshape_with_disjoint_set_test -- --random-seed=2862905288 -c1 -m2G --overprovisioned
Running 1 test case...
INFO  2023-02-09 18:37:07,753 seastar - Reactor backend: linux-aio
WARN  2023-02-09 18:37:07,753 [shard 0] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
INFO  2023-02-09 18:37:07,754 [shard 0] seastar - Created fair group io-queue-0, capacity rate 2147483:2147483, limit 12582912, rate 16777216 (factor 1), threshold 2000
INFO  2023-02-09 18:37:07,754 [shard 0] seastar - IO queue uses 0.75ms latency goal for device 0
INFO  2023-02-09 18:37:07,754 [shard 0] seastar - Created io group dev(0), length limit 4194304:4194304, rate 2147483647:2147483647
INFO  2023-02-09 18:37:07,754 [shard 0] seastar - Created io queue dev(0) capacities: 512:2000:2000 1024:3000:3000 2048:5000:5000 4096:9000:9000 8192:17000:17000 16384:33000:33000 32768:65000:65000 65536:129000:129000 131072:257000:257000
random-seed=2862905288

*** No errors detected

real	1m33,184s
user	0m0,990s
sys	0m1,985s
```

**Comment by xemul** [comment]
Cc @raphaelsc

**Comment by avikivity** [comment]
I remember seeing it a long time ago, tried pursuing it but lost the thread.

I guess we have a problem with aio wakeups.

**Comment by avikivity** [comment]
H have some vague memory of tracing it and seeing it did work.

**Comment by avikivity** [comment]
Ah, I remember now. As soon as you strace it, it works.

**Comment by xemul** [comment]
```
                if (idle_end - idle_start > _max_poll_time) {
                    ...
                    sleep();

```

it means it goes sleeping _every_ time it has nothing to do and `sleep()` is not cheap

**Comment by avikivity** [comment]
```
real	0m21.194s
user	0m0.394s
sys	0m0.544s
```

about 20 sec sleep time.

**Comment by avikivity** [comment]
```
INFO  2023-02-09 18:02:51,335 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:51,335 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:51,335 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:51,335 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,332 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,332 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:52,334 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,332 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,333 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,334 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,334 [shard 0] seastar - eventfd loop woke up
INFO  2023-02-09 18:02:53,334 [shard 0] seastar - eventfd loop woke up
```

So, there are 1-second pauses where it doesn't wake up.

**Comment by xemul** [comment]
It helps a little bit to all tests (in dev mode)

Now:
```
real	5m47,596s
user	44m1,066s
sys	12m16,782s
```

With `--idle-poll-time-us 200` to everyone:
```
real	5m3,639s
user	45m18,972s
sys	10m22,649s
```

**Comment by avikivity** [comment]
Aha, if I add a print, it starts working.

**Comment by avikivity** [comment]
Critical print is just before io_getevents.

**Comment by avikivity** [comment]
Great, doesn't reproduce in debug mode.

**Comment by xemul** [comment]
`epoll` backend also notably helps to twcs_reshape_with_disjoint_set_test, but still not perfect
* 1m33,184s -- default
* 0m18,197s -- epoll backend
* 0m1,688s -- idle-poll 200 
* 0m1,690s -- epoll & idle-poll 200

**Comment by raphaelsc** [comment]
> `epoll` backend also notably helps to twcs_reshape_with_disjoint_set_test, but still not perfect
> 
> * 1m33,184s -- default
> * 0m18,197s -- epoll backend
> * 0m1,688s -- idle-poll 200
> * 0m1,690s -- epoll & idle-poll 200

tried to reproduce it locally and the results are very inconsistent to me, epoll backend just took more than 1 min to me.

**Comment by avikivity** [comment]
It's some race. strace makes it go fast. a log before io_pgetevents makes it go fast.

**Comment by avikivity** [comment]
I looked at the eventfd integration in fs/aio.c, but it looks correct.

**Comment by avikivity** [comment]
I guess strace makes it slow so io_pgetevents() gets called after the IO completes.

**Comment by raphaelsc** [comment]
those outliers are beautiful, aren't they?!

```
**INFO  2023-02-09 14:49:24,332 [shard 0] seastar - slept for 965541us, sleep_count=4947, total_sleep=5903580us**
INFO  2023-02-09 14:49:24,333 [shard 0] seastar - slept for 124us, sleep_count=4948, total_sleep=5903704us
INFO  2023-02-09 14:49:24,335 [shard 0] seastar - slept for 14us, sleep_count=5000, total_sleep=5904585us
INFO  2023-02-09 14:49:24,338 [shard 0] seastar - slept for 25us, sleep_count=5100, total_sleep=5906137us
INFO  2023-02-09 14:49:24,342 [shard 0] seastar - slept for 18us, sleep_count=5200, total_sleep=5907677us
**INFO  2023-02-09 14:49:25,333 [shard 0] seastar - slept for 990690us, sleep_count=5224, total_sleep=6898728us**
INFO  2023-02-09 14:49:25,333 [shard 0] seastar - slept for 146us, sleep_count=5225, total_sleep=6898875us
INFO  2023-02-09 14:49:25,336 [shard 0] seastar - slept for 23us, sleep_count=5300, total_sleep=6900070us
INFO  2023-02-09 14:49:25,339 [shard 0] seastar - slept for 20us, sleep_count=5400, total_sleep=6901613us
**INFO  2023-02-09 14:49:26,334 [shard 0] seastar - slept for 993400us, sleep_count=5424, total_sleep=7895359us**
INFO  2023-02-09 14:49:26,334 [shard 0] seastar - slept for 183us, sleep_count=5425, total_sleep=7895543us
INFO  2023-02-09 14:49:26,335 [shard 0] seastar - slept for 137us, sleep_count=5436, total_sleep=7895899us
INFO  2023-02-09 14:49:26,337 [shard 0] seastar - slept for 20us, sleep_count=5500, total_sleep=7896897us
INFO  2023-02-09 14:49:26,340 [shard 0] seastar - slept for 24us, sleep_count=5600, total_sleep=7898450us
**INFO  2023-02-09 14:49:27,334 [shard 0] seastar - slept for 990901us, sleep_count=5679, total_sleep=8890569us**
INFO  2023-02-09 14:49:27,334 [shard 0] seastar - slept for 158us, sleep_count=5680, total_sleep=8890727us
```

**Comment by avikivity** [comment]
It's probably some 1-second timer that wakes us up. Without it we'd just sleep forever.

**Comment by avikivity** [comment]
Let's see if we go to sleep with a pending aio.

**Comment by raphaelsc** [comment]
yes, the pattern is clear.

**Comment by avikivity** [comment]
Very strange - we do not go to sleep with pending aio. Or I coded the check incorrectly.

**Comment by avikivity** [comment]
Let's see what io_uring has to say.

**Comment by avikivity** [comment]
reproduces!

**Comment by raphaelsc** [comment]
interestingly, not disarming the task quota time before going to sleep, allows for quick wakeup. but it's just hiding the dirt under the carpet. the problem is still there

**Comment by avikivity** [comment]
Yes it's just an earlier timer. But it's not a wakeup problem, we're going to sleep when we shouldn't.

**Comment by avikivity** [comment]
```c++
void
reactor::run_some_tasks() {
    if (!have_more_tasks()) {
        return;
    }
```

so  if have_more_tasks() lies, we'll deadlock.

**Comment by avikivity** [comment]
Maybe it's one of the enter_interrupt_mode()s.

**Comment by avikivity** [comment]
suspecting the syscall thread. This is a rename intensive workload.

**Comment by xemul** [comment]
> > `epoll` backend also notably helps to twcs_reshape_with_disjoint_set_test, but still not perfect
> > 
> > * 1m33,184s -- default
> > * 0m18,197s -- epoll backend
> > * 0m1,688s -- idle-poll 200
> > * 0m1,690s -- epoll & idle-poll 200
> 
> tried to reproduce it locally and the results are very inconsistent to me, epoll backend just took more than 1 min to me.

The `--random-seed` is important ... to some extent. Each run takes pretty different amount of time depending on the seed.

**Comment by avikivity** [comment]
```c++
@@ -2838,10 +2838,13 @@ class reactor::syscall_pollfn final : public reactor::pollfn {
     }
     virtual bool pure_poll() override final {
         return poll(); // actually performs work, but triggers no user continuations, so okay
     }
     virtual bool try_enter_interrupt_mode() override {
+        if (_r._thread_pool->has_pending_work()) {
+            return false;
+        }
```

makes it work. so syscall_pollfn::try_enter_interrupt_mode() is broken somehow.

**Comment by raphaelsc** [comment]
> ```c++
> @@ -2838,10 +2838,13 @@ class reactor::syscall_pollfn final : public reactor::pollfn {
>      }
>      virtual bool pure_poll() override final {
>          return poll(); // actually performs work, but triggers no user continuations, so okay
>      }
>      virtual bool try_enter_interrupt_mode() override {
> +        if (_r._thread_pool->has_pending_work()) {
> +            return false;
> +        }
> ```
> 
> makes it work. so syscall_pollfn::try_enter_interrupt_mode() is broken somehow.

isn't it the problem that reactor::syscall_pollfn::poll() only works on completed queue ("contains the requests already handled by the syscall thread")? completed work allows for new work in pending to be submitted. so we shouldn't enter interrupt mode if there's work in the pending queue.

**Comment by raphaelsc** [comment]
however, if we completed some work (opening space for new), we shouldn't have attempted to enter interrupt mode in the first place. So there's something awkward indeed.

**Comment by avikivity** [comment]
Why not enter interrupt mode? We can't do anything in the reactor thread until the syscall is processed and the result is in the completed queue.

**Comment by avikivity** [comment]
`taskset 1` "fixes" the problem

**Comment by avikivity** [comment]
If I add std::atomic_thread_fence() in thread_pool::work(), just before we load _main_thread_idle, it works.

**Comment by avikivity** [comment]
Preparing patch.

**Comment by raphaelsc** [comment]
> `taskset 1` "fixes" the problem

It surprised me that thread worker can be assigned to a different cpu.

**Comment by avikivity** [comment]
> > `taskset 1` "fixes" the problem
> 
> It surprised me that thread worker can be assigned to a different cpu.

Only when --overprovisioned.

**Comment by avikivity** [comment]
There's another bug there, but it's very minor compared to this one, let's get it merged first.

**Comment by xemul** [comment]
Fixed by 4f5a460db91de59018815106abb54c1246bbb014
```
    Update seastar submodule
    
    * seastar 943c09f869...9b6e181e42 (34):
...
      > thread_pool: prevent missed wakeup when the reactor goes to sleep in parallel with a syscall completion
...
```
